src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_controlnet.py

```diff
@@ -0,0 +1,695 @@
+# Copyright 2023 The HuggingFace Team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import inspect
+from typing import Any, Callable, Dict, List, Optional, Union
+
+import numpy as np
+import PIL.Image
+import torch
+from transformers import CLIPFeatureExtractor, CLIPTextModel, CLIPTokenizer
+
+from ...models import AutoencoderKL, ControlNetModel, UNet2DConditionModel
+from ...schedulers import KarrasDiffusionSchedulers
+from ...utils import (
+    PIL_INTERPOLATION,
+    is_accelerate_available,
+    is_accelerate_version,
+    logging,
+    randn_tensor,
+    replace_example_docstring,
+)
+from ..pipeline_utils import DiffusionPipeline
+from . import StableDiffusionPipelineOutput
+from .safety_checker import StableDiffusionSafetyChecker
+
+
+logger = logging.get_logger(__name__)  # pylint: disable=invalid-name
+
+
+def preprocess(image, width, height):
+    if isinstance(image, torch.Tensor):
+        return image
+    elif isinstance(image, PIL.Image.Image):
+        image = [image]
+
+    if isinstance(image[0], PIL.Image.Image):
+        image = [np.array(i.resize((width, height), resample=PIL_INTERPOLATION["lanczos"]))[None, :] for i in image]
+        image = np.concatenate(image, axis=0)
+        image = np.array(image).astype(np.float32) / 255.0
+        image = image[:, :, :, ::-1]  # RGB -> BGR
+        image = image.transpose(0, 3, 1, 2)
+        image = torch.from_numpy(image.copy())  # copy: ::-1 workaround
```

Comment:
@takuma104 super sorry, I'm not sure I follow what your change is doing here? https://github.com/huggingface/diffusers/pull/2407/commits/0f70cf573b2374d08bbea43e79b8ad28810a70f8

Reply:
Looking at the reference implementation ([example](https://github.com/lllyasviel/ControlNet/blob/main/gradio_normal2image.py#L34
)), it is apparent that the image preprocessing before conversion to torch.Tensor is performed using OpenCV. OpenCV generally uses the BGR color channel order, while PIL.Image commonly uses RGB. This is the conversion that needs to be made.

Yesterday, [I tried a quick operational check with all eight models](https://colab.research.google.com/drive/1AiR7Q-sBqO88NCyswpfiuwXZc7DfMyKA?usp=sharing#scrollTo=EN76qLrInJgi0), and I noticed that the behavior of the "normal" model was particularly strange. I think you'll notice the difference if you try it without the BGR to RGB conversion.

Reply:
hm, ok so I think this is because the normal model conditions on BGR channel order, not because of opencv

of the controlnet models, 

canny, depth, scribble, hed, and hough are all conditioned on grayscale images. for converting grayscale to RGB, the grayscale channel value is just repeated for each color channel. that means the RGB and BGR images values are equivalent.

the normal, pose, and seg models are all conditioned on colored images. The pose and seg models are conditioned on RGB channel order. The normal model is conditioned on BGR channel order. The normal model isn't conditioned on BGR channel order because the preprocessing is done with opencv but because the channel order is manually flipped here

https://github.com/lllyasviel/ControlNet/blob/1234076190105dc6541de3bc2aa7b3d24f8e3469/gradio_normal2image.py#L36

LMK if that make sense with your understanding, and I'll see if I can check with the original author since we should make sure this is right :) 

If this is right, I think we should choose whether or not to flip the channel order based on config on the serialized model -- it's relatively straightforward how to do this in diffusers but is a bit undocumented so I can take care of it.

Reply:
confirmed with author that the intention if for the conditioning to be on rgb channel order by default, they're going to check on what the intended channel order for the normal model is. So I'm going to change to the rgb ordering and if we need to add config on the pipeline to flip the channel ordering we can.

going to leave comment unresolved until we get confirmation on the normal model